content/en/tutorials/gitlab_ci_testcontainers/index.md

```diff
@@ -0,0 +1,432 @@
+  script:
+    - echo "Running build..."
+    - mvn clean package -DskipTests
+  artifacts:
+    paths:
+      - target/
+
+test_job:
+  stage: test
+  services:
+    - docker:26.1.2-dind
+  variables:
+    DOCKER_HOST: tcp://docker:2375
+    DOCKER_TLS_CERTDIR: ""
+    DOCKER_DRIVER: overlay2
+    LOCALSTACK_AUTH_TOKEN: $LOCALSTACK_AUTH_TOKEN
+
+  script:
+    - echo "Running tests..."
+    - mvn test
+
+  allow_failure: false
+
+```
+
+- `image: ubuntu:latest` - This specifies the base Docker image used for all jobs in the pipeline. `ubuntu:latest` is a popular and
+easy choice because it's a well-known, stable, and widely-supported Linux distribution. It ensures a consistent environment across
+all pipeline stages. Each job can define its own image (for example `maven` or `docker` images), but in this case a generic image with the 
+necessary dependencies (curl, Java, maven, docker) installed covers the needs for both stages.
+- `before_script` - these commands are run before any job script in the pipeline, on top of the Ubuntu image.
+- The two stages are defined at the top: `build` and `test`.
+- `cache` - caches the Maven dependencies to speed up subsequent pipeline runs. 
+- `.m2/repository` - this is the default location where Maven stores its local repository of dependencies.
+- The `script` section - specifies the scripts that run for each job.
+- `artifacts` - specifies the build artifacts (e.g., JAR files) to be preserved and passed to the next stages (the `target` folder).
+- The build job runs only on the `main` branch.
+- `docker:26.1.2-dind` - specifies the service necessary to use Docker-in-Docker to run Docker commands inside the pipeline job. This is 
+useful for integration testing with Docker containers.
+- Variables:
+  - `DOCKER_HOST: tcp://docker:2375` - sets the Docker host to communicate with the Docker daemon inside the dind service.
+  - `DOCKER_TLS_CERTDIR: ""` - we'll disable TLS to simplify the setup in a testing environment.
+  - `DOCKER_DRIVER: overlay2` - specifies the storage driver for Docker, ensuring better performance and compatibility.
+- The last line ensures that the pipeline fails if the tests fail.
+
+#### Executors
+
+We mentioned in the beginning that each job runs in a newly provisioned VM. You can also notice that the pipeline configuration mentions 
+a docker image, which is a template that contains instructions for creating a container. This might look confusing, but a runner is responsible 
+for the execution of one job. This runner is installed on a machine and implements 
+a certain [executor](https://docs.gitlab.com/runner/executors/). The executor determines the environment in which the job runs. By
+default, the GitLab-managed runners use a Docker Machine executor. Some other available executor options are: SSH, Shell, Parallels,
+VirtualBox, Docker, Docker Autoscaler, Kubernetes.
+
+Sometimes visualising the components of a pipeline can be tricky, so let's simplify this into a diagram:
```

Comment:
```suggestion
Sometimes visualising the components of a pipeline can be tricky, so let's simplify this into a diagram:
```

Just noticed that you used British spelling, I think we are trying to write everything in US English. @HarshCasper please confirm/deny.  :)

Reply:
fixed